### PR TITLE
Open browser with layout and styles

### PR DESCRIPTION
**Pull request by mcrumm**
I made two small tweaks to the initial implementation. The first is obviously to implement relative path rewriting for stylesheets. The second is to use the full page html when we get the root LiveView and inject the updated stylesheet tags. This gets us the layout as well.

I'd been toying with this for a minute, but in light of #1279 I definitely wanted to get some feedback :)

Before and after from `mix phx.new --live`:

2020-12-18 Edit: Incorporating @leandrocp changes in #1281 we can rewrite the relative image paths, too :)

|**Before**|**After**|
| :---: | :---: |
| screenshot |  screenshot |

// @leandrocp @mveytsman

### CODE REVIEW

File: lib/phoenix_live_view/test/client_proxy.ex [hunk 1]
```diff
@@ -85,6 +85,12 @@ defmodule Phoenix.LiveViewTest.ClientProxy do
       topic: Phoenix.LiveView.Utils.random_id()
     }
 
+    # We build an absolute path to any relative
+    # static assets through the root LiveView's endpoint.
+    static_path =
+      endpoint.config(:otp_app)
+      |> Application.app_dir("priv/static")
```
Comment (mcrumm) [inline lib/phoenix_live_view/test/client_proxy.ex:92]: I expect this may be a bit controversial. Very open to suggestions.

**Comment by leandrocp** [comment]
Hi @mcrumm thanks for working on this.

One benefit of `static_url: [path: "#{File.cwd!}/priv/static"]` to keep in mind is that it will resolve the path and load _all_ assets, including the images (the Phoenix logo in your example).

**Comment by mveytsman** [comment]
Thinking about it more, having the test endpoint give an absolute file path seems hacky, and like it might break some other tests people may have.

What if instead of rewriting stylesheets we inject `<base href="file://#{static_path}">` element into the head? This will have the effect of rewriting *all* relative URLs so stylesheets and images will all load correctly. Links will break -- but I think that since relative links won't work in the context of `open_browser` anyway this is acceptable.

What do you think @leandrocp @mcrumm 

By the way, if you check out a bare project and run tests, will assets be compiled into `priv/static` before tests are run? Or are we just assuming they are there because the user ran a development server?

**Comment by josevalim** [comment]
> <base href="file://#{static_path}">

Oh wow, I actually didn't know this was a thing. If it works, then it should considerably simplify this PR?

> By the way, if you check out a bare project and run tests, will assets be compiled into priv/static before tests are run? Or are we just assuming they are there because the user ran a development server?

You are correct, this is not guaranteed by I think it is a reasonably safe assumption.

**Comment by leandrocp** [comment]
> should considerably simplify this PR?

Yeah I think so :)

**Comment by mcrumm** [comment]
@josevalim since it doesn't look like we can rely on `<base>` (see #1281), this is the least amount of branching i could manage without additional input :)

wdyt?

**Comment by leandrocp** [comment]
> > At that point the html in the state should be resolved, right?
> 
> Correct. I incorporated your attrs check into #1280, which should produce more expected output for a single element :) wdyt?

@mcrumm  just to make sure we're on the same page :)

```elixir
    test "render view", %{live: view, open_fun: open_fun} do
      assert view |> open_browser() == view
    end
```

will render the whole page:

Screen Shot 2020-12-18 at 12 39 55 PM

```elixir
    test "render element", %{live: view, open_fun: open_fun} do
      element = element(view, "#scoped-render")
      assert element |> open_browser() == element
    end
```

will render only the html of that element:

Screen Shot 2020-12-18 at 12 40 16 PM

If I'm not mistaken that happens because the `html` returned by `render_tree(view_or_element)` has no `data-phx-main` attr, which makes me believe we don't even need that `render_tree`, unless it's necessary to sync/update the `state.html`. Otherwise that would be enough:

```
    {html, static_path} = call(view_or_element, :html)
    
    html
    |> Floki.traverse_and_update(...)
    |> write_tmp_html_file()
    |> open_fun.()
```

Wdyt?

**Comment by mcrumm** [comment]
> we don't even need that render_tree

If we want to render the whole page even when invoking `open_browser/1` with an Element, then you are correct :)

> display the whole page to help with the debugging

I would argue that if I'm consciously opening _just an element_ in the browser, I want to see the element in isolation, albeit with stylesheets loaded too so I can do some visual debugging.

**Comment by leandrocp** [comment]
> > we don't even need that render_tree
> 
> If we want to render the whole page even when invoking `open_browser/1` with an Element, then you are correct :)
> 
> > display the whole page to help with the debugging
> 
> I would argue that if I'm consciously opening _just an element_ in the browser, I want to see the element in isolation, albeit with stylesheets loaded too so I can do some visual debugging.

Yeah makes sense :)

Btw just tested this branch on a real project and it worked perfectly!

File: lib/phoenix_live_view/test/client_proxy.ex (hunk 1 above)
Comment (josevalim) [reply lib/phoenix_live_view/test/client_proxy.ex:92]: I think ti is fine for this PR. If someone requests, we can make it configurable.

**Review by josevalim — APPROVED** [review]

**Comment by mveytsman** [comment]
This is fantastic thanks @mcrumm @leandrocp !!

I'm bummed the `<base>` thing didn't actually work in practice. I looked into it a little more, I think the problem is that something like `/static/css/app.css` is not **relative** but actually **root-relative**. So it becomes relative to the root of the base url which in the case of file urls is just `/`. Doesn't seem to be any way around it :( but at least I learned something today :)

**Pull request merged**